cli/src/feature.rs

```diff
@@ -539,99 +656,119 @@ fn feature_set_stats(rpc_client: &RpcClient) -> Result<FeatureSetStats, ClientEr
+                    if stake_percent >= 0.001 || rpc_percent >= 0.001 {
+                        Some((
+                            cluster_config,
+                            ClusterInfoStatsEntry {
+                                stake_percent,
+                                rpc_percent,
+                            },
+                        ))
+                    } else {
+                        None
+                    }
                 },
-            )| {
-                let stake_percent = (stake as f64 / total_active_stake as f64) * 100.;
-                let rpc_nodes_percent = (rpc_nodes_count as f32 / total_rpc_nodes as f32) * 100.;
-                let mut software_versions = software_versions.into_iter().collect::<Vec<_>>();
-                software_versions.sort();
-                if stake_percent >= 0.001 || rpc_nodes_percent >= 0.001 {
-                    Some((
-                        feature_set,
-                        FeatureSetStatsEntry {
-                            stake_percent,
-                            rpc_nodes_percent,
-                            software_versions,
-                        },
-                    ))
-                } else {
-                    None
-                }
-            },
-        )
-        .collect())
+            )
+            .collect(),
+    })
 }
 
 // Feature activation is only allowed when 95% of the active stake is on the current feature set
 fn feature_activation_allowed(
     rpc_client: &RpcClient,
     quiet: bool,
-) -> Result<(bool, Option<CliClusterFeatureSets>), ClientError> {
-    let my_feature_set = solana_version::Version::default().feature_set;
-
-    let feature_set_stats = feature_set_stats(rpc_client)?;
-
+) -> Result<
+    (
+        bool,
+        Option<CliClusterFeatureSets>,
+        Option<CliClusterSoftwareVersions>,
+    ),
+    ClientError,
+> {
+    let cluster_info_stats = cluster_info_stats(rpc_client)?;
+    let feature_set_stats = cluster_info_stats.aggregate_by_feature_set();
+
+    let tool_version = solana_version::Version::default();
+    let tool_feature_set = tool_version.feature_set;
+    let tool_software_version = CliVersion(Some(semver::Version::new(
+        tool_version.major as u64,
+        tool_version.minor as u64,
+        tool_version.patch as u64,
+    )));
     let (stake_allowed, rpc_allowed) = feature_set_stats
-        .get(&my_feature_set)
+        .get(&tool_feature_set)
         .map(
             |FeatureSetStatsEntry {
                  stake_percent,
                  rpc_nodes_percent,
                  ..
              }| (*stake_percent >= 95., *rpc_nodes_percent >= 95.),
         )
-        .unwrap_or((false, false));
+        .unwrap_or_default();
+
+    let cluster_software_versions = if quiet {
+        None
+    } else {
+        let mut software_versions: Vec<_> = cluster_info_stats
+            .aggregate_by_software_version()
```

Comment:
It's a pity to have to do a 2nd loop through `cluster_info_stats` for this. Could we grab both aggregations on L720 above, and then only map to CliSoftwareVersionStats/sort/etc if `!quiet`?
I know it's tidy to have the two parallel aggregate functions...

Reply:
Are you concerned about perf here? I don't think this is a perf sensitive context personally. I split them up because I imagine there may be cases where we want to show the breakdown aggregated in only one way.

Reply:
I mean, it's not very perf sensitive -- we already have a zillion loops through various data in this path. But all other things being equal, I'd prefer to save the time.
If you've got plans for extra controls/new params, all other things aren't equal ;) and feel free to go with it as is.

Reply:
Sounds good! I'd like to keep it as is for now but wouldn't be upset if it was consolidated later